uc-crux-llvm/src/UCCrux/LLVM/FullType/CrucibleType.hs

```diff
@@ -81,55 +99,155 @@ toCrucibleType proxy =
+
+translateModuleDefines ::
+  forall arch.
   ( ?lc :: TypeContext,
     ArchOk arch
   ) =>
-  proxy arch ->
-  Ctx.Assignment CrucibleTypes.TypeRepr crucibleTypes ->
-  Ctx.Assignment (Const MemType) crucibleTypes ->
-  Either L.Ident (SomeAssign arch crucibleTypes)
-assignmentToFullType proxy crucibleTypes memTypes =
+  L.Module ->
+  ModuleTranslation arch ->
+  Either TypeTranslationError (TranslatedTypes arch)
+translateModuleDefines llvmModule trans =
   case makeModuleTypes ?lc of
-    Some moduleTypes ->
+    Some initialModuleTypes ->
+      let Identity (maybeResult, modTypes) =
+            runStateT
+              ( runExceptT
+                  ( mapM
+                      (translateDecl . LLVMTrans.declareFromDefine)
+                      (L.modDefines llvmModule)
+                  )
+              )
+              initialModuleTypes
+       in TranslatedTypes modTypes . DeclTypes . Map.fromList <$> maybeResult
+  where
+    translateDecl ::
+      L.Declare ->
+      ExceptT
+        TypeTranslationError
+        (StateT (ModuleTypes m) Identity)
+        (L.Symbol, FunctionTypes m arch)
+    translateDecl decl =
       do
-        (Some fullTypes, moduleTypes') <-
-          runStateT
-            ( Ctx.generateSomeM
-                (Ctx.sizeInt (Ctx.size crucibleTypes))
-                ( \idx ->
-                    case Ctx.intIndex idx (Ctx.size crucibleTypes) of
-                      Nothing -> panic "Impossible" ["Mismatched context size"]
-                      Just (Some idx') ->
-                        do
-                          Some fullTypeRepr <-
-                            toFullTypeM (getConst (memTypes Ctx.! idx'))
-                          pure $ Some fullTypeRepr
-                )
+        liftedDecl <-
+          case LLVMTrans.liftDeclare decl of
+            Left err -> throwError (BadLift err)
+            Right d -> pure d
+        Crucible.AnyCFG cfg <-
+          case Map.lookup (L.decName decl) (LLVMTrans.cfgMap trans) of
+            Just (_, anyCfg) -> pure anyCfg
+            Nothing -> throwError (NoCFG (L.decName decl))
+        let crucibleTypes = Crucible.cfgArgTypes cfg
+        let memTypes = fdArgTypes liftedDecl
+        let isVarArgs = fdVarArgs liftedDecl
+        unless
+          ( let numCrucibleTypes = Ctx.sizeInt (Ctx.size crucibleTypes)
+             in length memTypes == numCrucibleTypes
+                  || ( isVarArgs
+                         && length memTypes + 1 == numCrucibleTypes
+                     )
+          )
+          (throwError (BadLiftArgs (L.decName decl)))
+        Some fullTypes <-
+          Ctx.generateSomeM
+            ( Ctx.sizeInt (Ctx.size crucibleTypes)
+                - if isVarArgs
+                  then 1
+                  else 0
+            )
+            ( \idx ->
+                do
+                  Some fullTypeRepr <-
+                    withExceptT FullTypeTranslation $
+                      toFullTypeM (memTypes !! idx)
```

Comment:
Is this indexing always safe? Are the sizes of `memTypes` and `crucibleTypes` related?

Reply:
Yeah, they're asserted to be related in the `unless` above. I'll add a comment to that effect.